python_modules/dagster/dagster/_config/pythonic_config/config.py

```diff
@@ -346,29 +390,33 @@ def infer_schema_from_config_class(
     )
 
     fields: Dict[str, Field] = {}
-    for pydantic_field in model_cls.__fields__.values():
-        if pydantic_field.name not in fields_to_omit:
-            if isinstance(pydantic_field.default, Field):
+    for key, pydantic_field_info in model_fields(model_cls).items():
+        if _is_annotated_as_resource_type(
+            pydantic_field_info.annotation, pydantic_field_info.metadata
+        ):
+            continue
+
+        alias = pydantic_field_info.alias if pydantic_field_info.alias else key
+        if key not in fields_to_omit:
+            if isinstance(pydantic_field_info.default, Field):
                 raise DagsterInvalidDefinitionError(
                     "Using 'dagster.Field' is not supported within a Pythonic config or resource"
                     " definition. 'dagster.Field' should only be used in legacy Dagster config"
                     " schemas. Did you mean to use 'pydantic.Field' instead?"
                 )
 
             try:
-                fields[pydantic_field.alias] = _convert_pydantic_field(
-                    pydantic_field,
-                )
+                fields[alias] = _convert_pydantic_field(pydantic_field_info)
             except DagsterInvalidConfigDefinitionError as e:
                 raise DagsterInvalidPythonicConfigDefinitionError(
                     config_class=model_cls,
-                    field_name=pydantic_field.name,
+                    field_name=key,
                     invalid_type=e.current_value,
                     is_resource=model_cls is not None
                     and safe_is_subclass(model_cls, ConfigurableResourceFactory),
                 )
 
-    shape_cls = Permissive if model_cls.__config__.extra == Extra.allow else Shape
+    shape_cls = Permissive if model_config(model_cls).get("extra") == "allow" else Shape
```

Comment:
is there a first-class way of doing this in pydantic 2?

Reply:
It can be extracted directly (outside of the compat layer), but still relies on a `.get` call (it's an unstructured dictionary):

```python
>>> MyModel2(a_str='foo').model_config
{'extra': 'allow'}
```